jxl/src/frame/modular/transforms/squeeze.rs

```diff
@@ -133,45 +173,228 @@ fn smooth_tendency(b: i64, a: i64, n: i64) -> i64 {
+    if has_tail {
+        debug_assert!(in_avg.size().0 == w + 1);
+        debug_assert!(out.size().0 == 2 * w + 1);
+    }
+
+    let mut out_rect = out.as_rect_mut();
+    let mask = !(lanes - 1);
+    let scalar_start = if w >= lanes { h & mask } else { y_start };
+
+    let mut buf = [0f32; 512];
+    for y in (y_start..scalar_start).step_by(lanes) {
+        for dy in 0..lanes {
+            buf[dy] = f32::from_bits(in_avg.row(y + dy)[0] as u32);
+            buf[lanes + dy] = f32::from_bits(in_res.row(y + dy)[0] as u32);
+        }
+        let mut avg_first = D::F32Vec::load(d, &buf).bitcast_to_i32();
+        let mut res_first = D::F32Vec::load(d, &buf[lanes..]).bitcast_to_i32();
+
+        let mut prev_b = match out_prev {
+            None => avg_first,
+            Some(mc) => {
+                let mc_w = mc.data.size().0;
+                let mc = mc.data.as_rect();
+                for (dy, out) in buf[..lanes].iter_mut().enumerate() {
+                    *out = f32::from_bits(mc.row(y + dy)[mc_w - 1] as u32);
+                }
+                D::F32Vec::load(d, &buf).bitcast_to_i32()
+            }
+        };
+
+        let remainder_start = ((w - 1) & mask) + 1;
+        let remainder_count = w - remainder_start;
+        for x in (1..remainder_start).step_by(lanes) {
+            for dy in 0..lanes {
+                let avg_row = in_avg.row(y + dy);
+                let res_row = in_res.row(y + dy);
+                for dx in 0..lanes {
+                    buf[dx + lanes * 2 * dy] = f32::from_bits(avg_row[x + dx] as u32);
+                    buf[dx + lanes * (2 * dy + 1)] = f32::from_bits(res_row[x + dx] as u32);
+                }
+            }
+            let buf_arr = D::F32Vec::make_array_slice_mut(&mut buf);
+            D::F32Vec::transpose_square(d, buf_arr, 2);
+            D::F32Vec::transpose_square(d, &mut buf_arr[1..], 2);
+
+            for idx in 0..lanes {
+                let avg_next = D::F32Vec::load_array(d, &buf_arr[2 * idx]).bitcast_to_i32();
+                let res_next = D::F32Vec::load_array(d, &buf_arr[2 * idx + 1]).bitcast_to_i32();
+                let (a, b) = unsqueeze_impl(d, avg_first, res_first, avg_next, prev_b);
+                a.bitcast_to_f32().store_array(&mut buf_arr[2 * idx]);
+                b.bitcast_to_f32().store_array(&mut buf_arr[2 * idx + 1]);
+                avg_first = avg_next;
+                res_first = res_next;
+                prev_b = b;
+            }
+
+            D::F32Vec::transpose_square(d, buf_arr, 1);
+            D::F32Vec::transpose_square(d, &mut buf_arr[lanes..], 1);
+            for dy in 0..lanes {
+                let out_row = &mut out_rect.row(y + dy)[2 * x - 2..][..2 * lanes];
+                for group in 0..2 {
+                    let v = D::F32Vec::load_array(d, &buf_arr[dy + group * lanes]).bitcast_to_i32();
+                    v.store(&mut out_row[group * lanes..]);
+                }
+            }
```

Comment:
This emits code that copies to stack buffer, couldn't figure out how to remove those...

Reply:
I think it's fine :-)